packages/expo-cli/src/commands/upload/submission-service/archive-source/ArchiveFileSource.ts

```diff
@@ -0,0 +1,295 @@
+
+import { Platform } from '@expo/config';
+import { StandaloneBuild } from '@expo/xdl';
+import validator from 'validator';
+
+import { downloadAppArchiveAsync, uploadAppArchiveAsync } from '../utils/files';
+import { getAppConfig } from '../utils/config';
+import { existingFile } from '../../../../validators';
+import prompt from '../../../../prompt';
+import { SubmissionMode } from '../types';
+import log from '../../../../log';
+
+enum ArchiveFileSourceType {
+  url,
+  latest,
+  path,
+  buildId,
+  prompt,
+}
+
+interface ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType;
+  projectDir: string;
+  platform: Platform;
+}
+
+interface ArchiveFileUrlSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.url;
+  url: string;
+}
+
+interface ArchiveFileLatestSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.latest;
+}
+
+interface ArchiveFilePathSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.path;
+  path: string;
+}
+
+interface ArchiveFileBuildIdSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.buildId;
+  id: string;
+}
+
+interface ArchiveFilePromptSource extends ArchiveFileSourceBase {
+  sourceType: ArchiveFileSourceType.prompt;
+  projectDir: string;
+}
+
+export type ArchiveFileSource =
+  | ArchiveFileUrlSource
+  | ArchiveFileLatestSource
+  | ArchiveFilePathSource
+  | ArchiveFileBuildIdSource
+  | ArchiveFilePromptSource;
+
+async function getArchiveFileLocationAsync(
+  mode: SubmissionMode,
+  source: ArchiveFileSource
+): Promise<string> {
+  switch (source.sourceType) {
+    case ArchiveFileSourceType.prompt:
+      return await handlePromptSourceAsync(mode, source);
+    case ArchiveFileSourceType.url: {
+      const url = await handleUrlSourceAsync(mode, source);
+      return await getArchiveLocationForUrlAsync(mode, url);
+    }
+    case ArchiveFileSourceType.latest: {
+      const url = await handleLatestSourceAsync(mode, source);
+      return await getArchiveLocationForUrlAsync(mode, url);
+    }
+    case ArchiveFileSourceType.path: {
+      const path = await handlePathSourceAsync(mode, source);
+      return getArchiveLocationForPathAsync(mode, path);
+    }
+    case ArchiveFileSourceType.buildId: {
+      const url = await handleBuildIdSourceAsync(mode, source);
+      return await getArchiveLocationForUrlAsync(mode, url);
+    }
+  }
+}
+
+async function getArchiveLocationForUrlAsync(mode: SubmissionMode, url: string): Promise<string> {
+  if (mode === SubmissionMode.online) {
+    return url;
+  } else {
+    const tmpPath = pathJoin(os.tmpdir(), pathBasename(url));
+    log('Downloading your app archive');
+    return downloadAppArchiveAsync(url, tmpPath);
+  }
+}
+
+async function getArchiveLocationForPathAsync(mode: SubmissionMode, path: string): Promise<string> {
+  if (mode === SubmissionMode.online) {
+    log('Uploading your app archive to the temporary storage');
```

Comment:
```suggestion
    log('Uploading your app archive to the Expo Submission Service');
```

maybe this is a bit more clear. also might be nice to use `ora` here for a loading indicator beside the text, not sure if it plays nicely with having a progress bar though.

<img width="653" alt="Screen Shot 2020-05-24 at 8 14 56 PM" src="https://user-images.githubusercontent.com/90494/82775007-451b7180-9dfb-11ea-9085-44e0d9f200dc.png">

(example of ora loading indicator elsewhere in expo-cli)

Reply:
Yeah, I thought the same but actually if I recall correctly - it doesn't play nicely with a progress bar.